test/e2e/query_frontend_test.go

```diff
@@ -0,0 +1,236 @@
+		)
+
+		testutil.Ok(t, q.WaitSumMetricsWithOptions(
+			e2e.Equals(queryTimes+1),
+			[]string{"http_requests_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "handler", "query"))),
+		)
+	})
+
+	t.Run("query frontend works for labels APIs", func(t *testing.T) {
+		// LabelNames and LabelValues API should still work via query frontend.
+		labelNames(t, ctx, queryFrontend.HTTPEndpoint(), timestamp.FromTime(now.Add(-time.Hour)), timestamp.FromTime(now.Add(time.Hour)), func(res []string) bool {
+			return len(res) > 0
+		})
+		testutil.Ok(t, q.WaitSumMetricsWithOptions(
+			e2e.Equals(1),
+			[]string{"http_requests_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "handler", "label_names"))),
+		)
+
+		labelValues(t, ctx, queryFrontend.HTTPEndpoint(), "instance", timestamp.FromTime(now.Add(-time.Hour)), timestamp.FromTime(now.Add(time.Hour)), func(res []string) bool {
+			return len(res) > 0
+		})
+		testutil.Ok(t, q.WaitSumMetricsWithOptions(
+			e2e.Equals(1),
+			[]string{"http_requests_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "handler", "label_values"))),
+		)
+	})
+
+	t.Run("query frontend works for range query and it can cache results", func(t *testing.T) {
+		rangeQuery(
+			t,
+			ctx,
+			queryFrontend.HTTPEndpoint(),
+			queryUpWithoutInstance,
+			timestamp.FromTime(now.Add(-time.Hour)),
+			timestamp.FromTime(now.Add(time.Hour)),
+			14,
+			promclient.QueryOptions{},
+			func(res model.Matrix) bool {
+				return len(res) > 0
+			},
+		)
+
+		testutil.Ok(t, queryFrontend.WaitSumMetricsWithOptions(
+			e2e.Equals(1),
+			[]string{"thanos_query_frontend_queries_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "op", "query_range"))),
+		)
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "cortex_cache_fetched_keys"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(0), "cortex_cache_hits"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_added_new_total"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_added_total"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_entries"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_gets_total"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_misses_total"))
+
+		// Query is only 2h so it won't be split.
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "cortex_frontend_split_queries_total"))
+
+		testutil.Ok(t, q.WaitSumMetricsWithOptions(
+			e2e.Equals(1),
+			[]string{"http_requests_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "handler", "query_range"))),
+		)
+	})
+
+	t.Run("same range query, cache hit.", func(t *testing.T) {
+		// Run the same range query again, the result can be retrieved from cache directly.
+		rangeQuery(
```

Comment:
can we check that other endpoints are pass through?

Reply:
I already tested `labelNames` and `labelValues` here https://github.com/thanos-io/thanos/blob/master/test/e2e/query_frontend_test.go#L99-L118. Is it not enough?